util/election/runner.go

```diff
@@ -0,0 +1,187 @@
+func NewRunner(id int64, cfg *RunnerConfig, tracker *MasterTracker, cancel context.CancelFunc, el MasterElection) *Runner {
+	fixupRunnerConfig(cfg)
+	return &Runner{
+		Cancel:   cancel,
+		id:       id,
+		cfg:      cfg,
+		tracker:  tracker,
+		election: el,
+	}
+}
+
+// Run performs the process of performing a continuous election. It runs continuously until
+// the context is canceled or an internal error is encountered.
+func (er *Runner) Run(ctx context.Context, pending chan<- Resignation) {
+	// Pause for a random interval so that if multiple instances start at the same
+	// time there is less of a thundering herd.
+	pause := rand.Int63n(er.cfg.PreElectionPause.Nanoseconds())
+	if err := util.SleepContext(ctx, time.Duration(pause)); err != nil {
+		return
+	}
+
+	glog.V(1).Infof("%d: start election-monitoring loop ", er.id)
+	if err := er.election.Start(ctx); err != nil {
+		glog.Errorf("%d: election.Start() failed: %v", er.id, err)
+		return
+	}
+	defer func(ctx context.Context, er *Runner) {
+		glog.Infof("%d: shutdown election-monitoring loop", er.id)
+		er.election.Close(ctx)
+	}(ctx, er)
+
+	for {
+		glog.V(1).Infof("%d: When I left you, I was but the learner", er.id)
+		if err := er.election.WaitForMastership(ctx); err != nil {
+			glog.Errorf("%d: er.election.WaitForMastership() failed: %v", er.id, err)
+			return
+		}
+		glog.V(1).Infof("%d: Now, I am the master", er.id)
+		er.tracker.Set(er.id, true)
+		masterSince := er.cfg.TimeSource.Now()
+
+		// While-master loop
+		for {
+			if err := util.SleepContext(ctx, er.cfg.MasterCheckInterval); err != nil {
+				glog.Infof("%d: termination requested", er.id)
+				return
+			}
+			master, err := er.election.IsMaster(ctx)
+			if err != nil {
+				glog.Errorf("%d: failed to check mastership status", er.id)
+				break
+			}
+			if !master {
+				glog.Errorf("%d: no longer the master!", er.id)
+				er.tracker.Set(er.id, false)
+				break
+			}
+			if er.ShouldResign(masterSince) {
+				glog.Infof("%d: queue up resignation of mastership", er.id)
+				er.tracker.Set(er.id, false)
+
+				done := make(chan bool)
+				r := Resignation{ID: er.id, er: er, done: done}
+				pending <- r
+				<-done // block until acted on
+				break  // no longer master
+			}
+		}
+	}
+}
+
+// ShouldResign randomly decides whether this runner should resign mastership.
+func (er *Runner) ShouldResign(masterSince time.Time) bool {
+	now := er.cfg.TimeSource.Now()
+	duration := now.Sub(masterSince)
+	if duration < er.cfg.MasterHoldInterval {
+		// Always hold onto mastership for a minimum interval to prevent churn.
+		return false
+	}
+	// Roll the bones.
+	odds := er.cfg.ResignOdds
+	if odds <= 0 {
+		return true
+	}
+	return rand.Intn(er.cfg.ResignOdds) == 0
+}
+
+// Resignation indicates that a master should explicitly resign mastership, by invoking
+// the Execute() method at a point where no master-related activity is ongoing.
+type Resignation struct {
+	ID   int64
+	er   *Runner
+	done chan<- bool
```

Comment:
Could make this `chan struct{}` and just close it in `Execute`.

Reply:
We could indeed, but that would contradict the spirit of this PR.

Reply:
SGTM, let's leave it as is.